Develop

ngtspc opened the pull request.

ngtspc committed:
first commit

ngtspc committed:
first commit

ninjavu commented on json_files/authors.json:
```diff
@@ -0,0 +1,43 @@
+[ 
```
What's behavior of your program in case of absence of your json file?

ngtspc replied:
The case has been taken into account

ninjavu commented on language.rb:
```diff
@@ -0,0 +1,70 @@
+    TEXTS[@@language]
+  end
+end
+
```
What's wrong? :) Why so many empty spaces? Be careful with styling of your code.

ngtspc replied:
Fixed. My bad

ninjavu commented on authors.rb:
```diff
@@ -0,0 +1,18 @@
+require_relative 'file_handler'
+
+class Authors < Rating
```
It's better to keep following files in folder like entities, to keep relevant files in separate namespaces is good idea

ngtspc replied:
Fixed

ninjavu commented on language.rb:
```diff
@@ -0,0 +1,70 @@
+    @@language = gets.chomp.to_sym
+    unless LANGUAGES.include?(@@language)
+      puts "Please choose ':en' or ':ru' "
+    @@language = gets.chomp.to_sym
```
spaces

ngtspc replied:
Fixed

ninjavu commented on library.rb:
```diff
@@ -0,0 +1,150 @@
+  include Language
+  attr_reader :authors_lib, :books_lib, :clients_lib, :orders_lib
+
+  def initialize(authors_lib = Authors.new,
```
It's better to have already preloaded arrays of classes.

ngtspc replied:
Done

ninjavu commented on library.rb:
```diff
@@ -0,0 +1,150 @@
+    end
+  end
+
+  def read_file(file_name)
```
it's not a responsibility of Library class

ninjavu commented on library.rb:
```diff
@@ -0,0 +1,150 @@
+    pp entity
+  end
+
+  def show_books
```
Better to move it to models [Author, Book...]

ngtspc replied:
Done

ninjavu commented on main.rb:
```diff
@@ -0,0 +1,4 @@
+require_relative 'library'
+
+library = Library.new
+library.run
```
spaces

ninjavu commented on rating.rb:
```diff
@@ -0,0 +1,25 @@
+class Rating
+  attr_reader :orders
+
+  def initialize(orders = FileHandler.new('orders').parse_file)
```
it's better to pass just structure, retrieve a data from FileHandler isn't best idea, better to execute it before and just pass structure to Rating

ngtspc replied:
Fixed

ninjavu commented on rating.rb:
```diff
@@ -0,0 +1,25 @@
+    @orders = orders
+  end
+
+  def top_ids(ntimes, orders_key)
```
ntimes -> n_times I ruby we are using snake case

ngtspc replied:
Fixed

ninjavu commented on library.rb:
```diff
@@ -0,0 +1,150 @@
+  end
+
+  def top_books(ntimes, orders_key)
+    books_lib.top_n(ntimes, orders_key, books_lib.books.parse_file, 'id') do |index, hash_top_ids|
```
good idea with block, but let's keep iterators in Rating class, because a Library shouldn't know something about Rating operations

ngtspc replied:
Fixed. The structure of classes has been changed

ninjavu reviewed (changes requested):
First of all, looks good. But I would recommend a few changes, some of them in the comments, some of them: 
1. I don't see Gemfile with rubocop, a lot of style errors, please, be careful with it.
2. I think you can be better with architecture, as we discussed before, it's better when each of classes should implement own logic and be lowly coupled with another classes. You have a lot of IRRELEVANT logic in Library class, which better to place in entities, Rating etc... Revise your architecture a little bit
3. You should think about errors and errors handling, e.g. what if I will put empty folder without json files, what if their names will be different, what if they will have incorrect fields etc... Files are not only 1 place to think about it
4. Please, read Ruby conventions, I still see variables like `ntimes` and it's bad.
5. Be careful with git commits messages, PR names, it's really important

ngtspc committed:
What was changed:
1. Files were transfered to folders
2. Classes were rewriten
3. Gemfile was added
4. The case if json file does not exist was taken into account

ngtspc committed:
What was changed:
1. Files were transfered to folders
2. Classes were rewriten
3. Gemfile was added
4. The case if json file does not exist was taken into account

ninjavu commented on json_files/clients.json:
```diff
@@ -0,0 +1,122 @@
+    "last_name": "sdtfyg",
+    "address": "dgrfg"
+  }
+]
```
Check your code with rubocop, newlines are missing

ngtspc replied:
Fixed

ninjavu commented on service/rating.rb:
```diff
@@ -0,0 +1,56 @@
+  end
+
+  def top_books(ntimes)
+    if @orders.instance_of?(Integer) || @orders.empty?
```
Why you paste it here? It should be in FileHandler and it should work for all files

ngtspc replied:
We discussed about it before and resolved this question

ninjavu commented on service/rating.rb:
```diff
@@ -0,0 +1,56 @@
+    end
+  end
+
+  def top_authors(ntimes)
```
It's code duplication, let's have 1 method instead 3

ngtspc replied:
Fixed

ninjavu commented on objects/client.rb:
```diff
@@ -0,0 +1,68 @@
+  end
+
+  def fill_info(first_name, last_name, address)
+    @client = {
```
Old syntax hash

ngtspc replied:
I think it's better to have an old syntax as we will push this hash to json. And as json understand only old suntax I do not see a point to convert hash from old to new and otherwise

ngtspc replied:
Fixed. My bad

ngtspc replied:
> Old syntax hash

Fixed. My bad

ninjavu commented on objects/client.rb:
```diff
@@ -0,0 +1,68 @@
+    @clients.find { |client| "#{client['first_name']} #{client['last_name']}" == @full_name }
+  end
+
+  def info
```
Please, refactor it, I mean integrate all conditions

ngtspc replied:
Please give me some more details regarding this comment. What do you mean by 'all conditions'?

ninjavu commented on service/rating.rb:
```diff
@@ -0,0 +1,56 @@
+    end
+  end
+
+  def top_clients(ntimes)
```
ntimes -> n_times, I said it before :)

ngtspc replied:
Fixed

ninjavu commented on service/language.rb:
```diff
@@ -0,0 +1,78 @@
+    @@language = gets.chomp.to_sym
+    unless LANGUAGES.include?(@@language)
+      puts 'Please choose ":en" or ":ru" '
+    @@language = gets.chomp.to_sym
```
spaces, RUBOCOP!!!

ngtspc replied:
Fixed

ngtspc committed:
Fir process

ngtspc committed:
Fir process

ngtspc committed:
Structure changing

ninjavu commented on Gemfile:
```diff
@@ -0,0 +1,6 @@
+
+source 'https://rubygems.org'
+
+# gem "rails"
```
remove unused gem

ninjavu commented on index.rb:
```diff
@@ -0,0 +1,54 @@
+  database
+end
+
+parse_file
```
new line is missing

ninjavu commented on index.rb:
```diff
@@ -0,0 +1,54 @@
+
+def parse_file
+  database = []
+  if File.exist?("./json_files/gavno.json")
```
what?

ninjavu commented on index.rb:
```diff
@@ -0,0 +1,54 @@
+      p database = create_file(n = gets.chomp.to_i)
+      write_file(database)
+    else
+      p 'не зашло'
```
What?

ngtspc committed:
index.rb has been deleted

ngtspc committed:
Main.rb has been changed

ngtspc committed:
Delete main.rb

ninjavu commented on service/rating.rb:
```diff
@@ -0,0 +1,47 @@
+          .keys
+  end
+
+  def top_books(_n_times, orders, books, top_ids)
```
Why _n_times with _? And seems you don't use it in function, more attention

ngtspc replied:
Fixed

ninjavu commented on service/rating.rb:
```diff
@@ -0,0 +1,47 @@
+    end
+  end
+
+  def top_authors(_n_times, orders, authors, top_ids)
```
The same regarding _n_times

ngtspc replied:
Fixed

ninjavu commented on service/rating.rb:
```diff
@@ -0,0 +1,47 @@
+    end
+  end
+
+  def top_clients(_n_times, orders, clients, top_ids)
```
The same regarding _n_times

ngtspc replied:
Fixed

ninjavu commented on service/rating.rb:
```diff
@@ -0,0 +1,47 @@
+      p 'Unfortunately, our database is empty at the moment.'
+    else
+      top_ids.each_with_index do |id, index|
+        hash_top_ids = clients.find { |object| object.id.to_i == id.to_i }
```
object? Be more clear with naming, e.g. client

ngtspc replied:
Fixed

ngtspc committed:
Merge branch 'develop' of https://github.com/ngtspc/library.oop into develop
Adding main.rb

ninjavu commented on service/rating.rb:
```diff
@@ -0,0 +1,47 @@
+# Description/Explanation of Rating module
+module Rating
+  def top_ids(n_times, orders, key_name)
+    orders.group_by { |object| object.instance_variable_get("@#{key_name}") }
```
It's nice that you've investigated `instance_variable_get`, but why do you use it here? It's better to avoid using following constructions for now.

ngtspc replied:
I've changed to object.send("#{key_name}"). Is it more acceptable?

ninjavu commented on service/rating.rb:
```diff
@@ -0,0 +1,47 @@
+
+require_relative 'file_handler'
+
+# Description/Explanation of Rating module
```
You can delete this comment and skip following warnings in rubocop configuration.

ngtspc replied:
Fixed

ninjavu commented on service/rating.rb:
```diff
@@ -0,0 +1,47 @@
+require_relative 'file_handler'
+
+# Description/Explanation of Rating module
+module Rating
```
Please, refactor it, currently you execute top_ids each time outside module and pass it in each function, it should be only inside

ngtspc replied:
Fixed

ninjavu commented on service/rating.rb:
```diff
@@ -0,0 +1,47 @@
+      p 'Unfortunately, our database is empty at the moment.'
+    else
+      top_ids.each_with_index do |id, index|
+        hash_top_ids = clients.find { |object| object.id.to_i == id.to_i }
```
result is not hash_top_ids, refactor it

ngtspc replied:
Fixed. I've changed it to top_author, top_client and so on. I think it is more acceptable as we get an entity of our top objects collection on each iteration. Please comment this moment so do I know if the naming of variable is correct

ngtspc committed:
Some changed in parsing files. Require a new module

ngtspc committed:
all text was transfered to module. Rating has now 1 top function

ngtspc committed:
Some fixies in functions

ngtspc committed:
Update order.rb

ngtspc committed:
Fixed spaces in hashes

ngtspc committed:
Some syntax update

ninjavu commented on service/rating.rb:
```diff
@@ -0,0 +1,22 @@
+require_relative 'file_handler'
+
+module Rating
```
Regrading `Rating module`, nice work! Thanks.

ninjavu commented on service/object_creator.rb:
```diff
@@ -0,0 +1,74 @@
+module ObjectCreator
```
Regarding ObjectCreator at all, I see that each of your conditions have pretty similar logic, let's think how could you improve your code with DRY principles?

ngtspc committed:
Optimisation of ObjectCreator module and classes which use it

ngtspc committed:
Changes in ObjectCreator (ObjectSetter) module and files which require it

ninjavu commented on service/object_setter.rb:
```diff
@@ -0,0 +1,13 @@
+module ObjectSetter
+  
+  def create_file(attributes)
```
choose better method name

ngtspc replied:
Fixed

ninjavu commented on service/rating.rb:
```diff
@@ -0,0 +1,22 @@
+
+  def top_n(n_times, orders, key_name, collection, key_name2)
+    if orders.empty?
+      p 'Unfortunately, this database is empty at the moment.'
```
Please  i18n

ngtspc replied:
Fixed

ninjavu commented on service/object_setter.rb:
```diff
@@ -0,0 +1,13 @@
+  def create_file(attributes)
+    result = {}
+    attributes.each do |field|
+      p "Please enter #{field}"
```
please i18n

ngtspc replied:
Fixed

ngtspc committed:
Delete object_creator.rb

ngtspc committed:
Fixed naming in ObjectSetter, service folder and optimized language module

ngtspc committed:
Merge branch 'develop' of https://github.com/ngtspc/library.oop into develop

ngtspc committed:
Delete file_handler.rb

ngtspc committed:
Delete language.rb

ngtspc committed:
Delete library_helper.rb

ngtspc committed:
Delete object_setter.rb

ngtspc committed:
Delete rating.rb

ninjavu commented on library.rb:
```diff
@@ -0,0 +1,177 @@
+                        DateTime.parse(object['created_at']),
+                        DateTime.parse(object['updated_at']),
+                        object['author_id'].to_i,
+                        object['price'].to_i)
+      @books.push(entity)
+    end
+    @books
+  end
+
+  def parse_authors
+    get_collections('authors') do |object|
+      entity = Author.new(object['first_name'],
+                          object['last_name'],
+                          object['book_id'].to_i)
+      @authors.push(entity)
+    end
+    @authors
+  end
+
+  def parse_clients
+    get_collections('clients') do |object|
+      entity = Client.new(object['id'].to_i,
+                          object['first_name'],
+                          object['last_name'],
+                          object['address'])
+      @clients.push(entity)
+    end
+    @clients
+  end
+
+  def parse_orders
+    get_collections('orders') do |object|
+      entity = Order.new(object['book_id'].to_i,
+                         DateTime.parse(object['created_at']),
+                         object['client_id'].to_i,
+                         object['payed'].to_i)
+      @orders.push(entity)
+    end
+    @orders
+  end
```
Notice that I've selected multiple lines and all methods should be in separate class (abstraction) like `class Storage` to interact with json files, like:

`Storage.get('books')`
`Storage.update...`
`Storage.delete...`

ngtspc replied:
Fixed

ninjavu commented on library.rb:
```diff
@@ -0,0 +1,177 @@
+    @orders
+  end
+
+  def load_files
```
Please remove it and call those methods directly in `initialize`

ngtspc replied:
Fixed

ninjavu commented on library.rb:
```diff
@@ -0,0 +1,177 @@
+    @orders = parse_orders
+  end
+
+  def client_info
```
Keep client related methods in Client class

ngtspc replied:
Fixed

ninjavu commented on library.rb:
```diff
@@ -0,0 +1,177 @@
+    client
+  end
+
+  def write_order(book, client)
```
Keep order related methods in Order class

ngtspc replied:
Fixed

ninjavu commented on library.rb:
```diff
@@ -0,0 +1,177 @@
+    puts phrases_list[:payment_accepted]
+  end
+
+  def buy_book
```
Keep Book related method in Book class

ngtspc replied:
Fixed

ninjavu commented on library.rb:
```diff
@@ -0,0 +1,177 @@
+  end
+
+
+  def add_entity(file_name, new_entity)
```
It also should place in `class Storage` that we've discussed

ngtspc replied:
Fixed

ninjavu commented on library.rb:
```diff
@@ -0,0 +1,177 @@
+  end
+
+  def run
+    parse_authors
+    parse_books
+    parse_clients
+    parse_orders
```
Delete it

ngtspc replied:
Fixed

ninjavu commented on library.rb:
```diff
@@ -0,0 +1,177 @@
+    FileHandler.new(file_name).write_file(file)
+  end
+
+  def run
```
Right concept of method run, but I would ask you to have some condition and just place corresponding action from corresponding class, e.g.

some condition -> ExampleClass.example_method and that's it

ngtspc committed:
Lubrary class function has been trasnfered to appripriate classes + some other fixes

ngtspc committed:
Merge branch 'develop' of https://github.com/ngtspc/library.oop into develop

ninjavu reviewed (approved):
Looks good @ngtspc ! Please add rubocop config and let's merge to master, I think you will find a places to do small refactoring in the future, but now let's mode on :)

ngtspc committed:
Added rubucop and deleted 'top_ids' in Library class
